### ISSUE TITLE
What in the world is a mine?

### ISSUE THREAD

**Issue by cereallarceny** [OpenMined/mine#1]
## What is a mine?
Mines are a data repository where people can upload personal information and begin contributing to neural networks on the Sonar blockchain.

## What are our requirements?
1. Mines should (in theory) be online and accessible 24/7/365 (or some reasonable level of "consistent" online access).
2. Mines should never be known to Sonar.
3. Mines should be 100% managed and owned by the user they're associated with.
4. Mines should act as a data repository as well as be able to run neural networks they download from the Sonar blockchain.  The neural network is run on the mine and will use only the data they are storing themselves.  Upon final computation, the generated gradient is uploaded back to Sonar and appended to the blockchain record.
5. (Optional) It should be free to run a mine, and in fact, you should be paid for it by participating in bounties.

## How do mines work?
With that said, there's a number of proposed ways (each with pro's and con's of their own), in which we could build these mines.  Here are the current working theories (in no particular order):

### 1. Cloud Portal
Users would have the ability to go to the Open Mined website and create an account.  From here, they would manage their mine.  This is akin to how a Dropbox account might work.  To the average person, this is a wonderful user experience and requires very little set up.  However, **the plot thickens...**

#### Storage
Data would be stored on a cloud service like AWS, Google Cloud, Azure, or a number of private cloud solutions.  The inherent problem here is that we cannot create an Amazon Web Services account (for instance) on behalf of a user.  One of two (undesirable) situations would unfold:

1. We ask them to create an AWS account, create an IAM user, and then input those credentials on the Open Mined portal.
2. We have a master Open Mined AWS account and we simply create a "section" for that user that we manage on their behalf.  There would need to be privacy controls to ensure that we don't have access to their data.

#### Privacy and Control
This may seem a bit antithetical to the mission of Open Mined, but I believe it may be a fair compromise, especially since users still "own their own information".  Some people may find it a bit hypocritical to claim that users can host their own data and in reality, it's all going to a major cloud provider.  Food for thought.

#### Good News
Development, albeit some hurdles listed above, would be relatively straight forward.  We'd spend the majority of it building a front-end to interact with S3, EC2, Lambda, and the like.  This would be great for providing a proof of concept, especially to developers who are likely to already have an AWS account.  However, it's lacking some inherent privacy concerns and has a relatively high barrier to entry.  This wouldn't be accessible to the "common person".

### 2. NAS
Users would purchase a network attached storage device and load an image on to it that would run in their homes 24/7.  For sake of argument and simplicity, let's assume this is a Raspberry Pi that comes with an image the Open Mined community has created.

#### Installation
In theory, this should include the following steps:
1. Unbox the Raspberry Pi
2. Plug the ethernet cable into the user's router and the power cable into a wall outlet
3. The image boots up and should be accessible from their computer connected to the same network within a few minutes.  They go to a web address: https://xxx.xxx.xxx.xxx and are prompted to log in to their Open Mined account.
4. Upon logging in, they're now able to begin uploading data or connecting social media accounts to download their profiles.  This part needs to be thought out more as well... potentially with the addition of an "adapter marketplace".

#### Cost
This puts the onus of fronting the money for the cost of the Raspberry Pi on the end-user.  In a perfect world, we would be able to sink this cost ourselves and have it offset by taking a cut of bounties, allowing the hardware to be paid for.  This could remove the barrier to entry for the "common person" but creates a massive business problem.  We either deal with that business problem or offload the cost to users.  For developers and early adopters, this probably isn't a big deal, but it's worth considering a more long-term solution going forward.

#### Development
I don't have an in depth knowledge of creating images for a Raspberry Pi.  I would love for someone more knowledgeable to bring their mind to this discussion.  It's my understanding though that this sort of development is non-trivial.  Essentially we would need to create a mining application that interacts with Sonar as well as the disk image behind our own NAS.

#### Good News
This solves an inherent issue with privacy and having true control over your information.  A user, at any point, could disconnect from the whole system with the pulling of a plug.  It also means that there's no hosting bill to tend to (even if we subsidize this against their collected bounties).  In many ways, this is an easier solution to the problem for our users but comes at some costs both financial and developmental.

## Contributing
This is intended to be a working document - let the discussion ensue!

**Comment by kevinahuber** [comment]
How does Sonar ensure the validity of a gradient without knowing any details of the mine?

Is the purpose of the mine having a consistent uptime for the mine owner? If so- doesn't it only need to be up as long as the mine owner wants to input data? 

Is there a certain amount of redundancy that is needed? If so, how would that be achieved with a raspberry pi?

If we are able to have a model where a user can host on a pi, could they also host on their own AWS account _without_ giving Open Mined IAM access? It is increasingly less non-developer accessible, but more data ownership can be given. A cloud formation template or something of the sort to set up your very own mine! What are all of the requirements to connect to a Sonar model?

**Comment by iamtrask** [comment]
- How does Sonar ensure the validity of a gradient without knowing any details of the mine?

Great point. To that end, I suppose Mines should be able to register with Sonar but it should be important that Sonar does not control the mines. Gradients are voluntarily pushed to Sonar as a "Hey! I'd like to participate in your open challenge to improve this model". Patrick and I previously discussed Sonar not knowing about Mines at all, but your point is well taken. The schema of available data needs to be shared to Sonar in order for companies to start competitions knowing that data is available.

- Is the purpose of the mine having a consistent uptime for the mine owner? If so- doesn't it only need to be up as long as the mine owner wants to input data?

Yes, it's to maximize revenue for the owner. In theory they could pull it down but as it's likely to take several days to train models, it's in everyone's interest for us to build Mines with the hope that they have very consistent uptime

- Is there a certain amount of redundancy that is needed? If so, how would that be achieved with a raspberry pi?

I heard about the raspberry pi idea through slack. It's a great idea and certainly compatible with the whole "you control your data" message that we want laymen to understand (who might be confused if they were re-uploading it to Amazon EC2 even though it's quite safe there). That being said, I don't think it's wise for us to couple our implementation to any particular hardware, especially this early on. Priority 1 at this point should be speed in an ideal environment (which a raspberry pi is not from a performance standpoint), architecture specific deployments is undoubtedly necessary but second to speed. 

- If we are able to have a model where a user can host on a pi, could they also host on their own AWS account without giving Open Mined IAM access? It is increasingly less non-developer accessible, but more data ownership can be given. A cloud formation template or something of the sort to set up your very own mine! What are all of the requirements to connect to a Sonar model?

Definitely. We should build the Mines from an architecture agnostic perspective first, and then think about optimizing for specific deployments. All that is to say, It's too early to worry about making installation easy. (We don't even know the constraints of that installation yet)

**Comment by cereallarceny** [comment]
> Is the purpose of the mine having a consistent uptime for the mine owner? If so- doesn't it only need to be up as long as the mine owner wants to input data?

To add to this as well, I think it would be natural for us to offer Open Mined data in the form of an oAuth protocol someday.  This is a long term play, but it's certainly a reality.  A web application could authenticate with Open Mined and pull data from their mine that they are permitted to pull.  Same could even go for write access... you get the idea.  @iamtrask's description of maximizing effort is true as well.  Just thinking out loud.

**Comment by anoff** [comment]
**cloud hosting**
Regarding (cloud) hosting I'd make it top prio to have the mine as a one-step installation. Possible solutions docker (more techy) or a bundled desktop app (@cereallarceny and myself should be able to pull that one off if we stick to compiled webapps). Docker should be super easy, the apps might take a big longer but def. no blocking issues there imo. A docker image should also make it relatively straight-forward to have a Open Mined AMI for ppl(devs) to run their Mine online. Syncing data might be another issue here..
How long do you expect the training for an individual Mine to take? I was under the assumption it's going to be maximum one hour. I get the feeling you guys assume waaayyy more. If it requires little time for the Mine to actually train it might be worth to have a central **Mine Hive**(?) (opt-in) where Mines are registered with the type of data they hold. Then we could notify Mine owners via e-mail, message etc. if there is a Model that is worth for them to be trained. Added benefit of this solution is that Farmers can see in advance how many datasets are available for their model.

**decentralised Mine**
Another addition to improve the availability of the mine while still keeping the data 💯 % with the user i.e. not uploading the unencrypted data might be decentralised Mine infrastructure. What I mean by that is that the user could have different Mine instances, e.g. one mobile, maybe a web-app based, hardware mine (raspi) and a desktop client. Kind of like wallets for other blockchain assets. The challenge in this scenario is to have all Mines synchronised regarding the data they hold. One approach I really like here is to have the data client-side encrypted and have it synced via a webservice that Open Mined offers their users. A nice example of this client-side enryption is [clipperz password manager](clipperz.is/app/)

Constraints:
* data should be rather small (<100MB) or needs to be split into separately syncable chunks
* mining of the model should work on multiple platforms (mobile?)
* mining shoud be quick enough so it doesn't require multi-day uptime (that would make cloud-hosted mines a high priority issue as most ppl won't have devices running 24/7)

Pro:
* basically opens the ecosystem for ppl to plug in additional devices
* makes data ingestion a lot easier because depending on the data source the Mine could be run directly where the data is generated (e.g. gps tracks on mobile phone, twitter on desktop, temperature(?) on raspi)
* keep 💯 % true to the "keep your data private" aspect
* no accounts to be generated by _users_

Cons:
* Constraints above
* infrastructure costs - however pure traffic is rather cheap and could be sponsored by Open Mined
* _might_ concern ppl about privacy (but far less than server hosted)

**Comment by iamtrask** [comment]
Something to remember about the Mine. It's not just where data is stored. It's also where models are trained on the data. In order for this process to happen, the data must be **decrypted**. Thus, any environment we use for a Mine has to be inherently safe.

**Comment by iamtrask** [comment]
Also, some food for thought on our priorities for Mine development.

1) The first “customers” we need to get onboard (to deploy it) are the actively contributing members of our team (us!). So, the first, most natural deployment, should be a script. We need to make development as easy as possible as early as possible.
2) The next most important customer (at this juncture) is people who are passionate about the project (but not necessarily actively developing it) but still want to try it (think “tech enthusiast”)… this party could use a free tier level Amazon EC2 instance… basically some automation and good instructions around (1) and they'll be happy
3) THEN we get to the real stuff… things like raspberry pi, iOS, Android (and maybe cloud). The goal here is to use that consumers already have as much as possible to make the barrier to entry as low as possible.

**Comment by kevinahuber** [comment]
Besides registering with Sonar models and uploading gradients, does the mine need to have internet connection?

**Comment by iamtrask** [comment]
Besides those two things, I don't believe so.

**Comment by kevinahuber** [comment]
What is the cost of pausing/restarting training? 
Is a purpose of the mine also to store the data itself for the user's access? Earlier descriptions of beginning phases makes me think not, but long term is this the goal? 

If access is not a high priority, then I would think bundled desktop apps and decentralized models are adapters on the initial mine. We start with something that has a single command deployment- I concur with @anoff - Docker is a solid starting point

**Comment by iamtrask** [comment]
The cost of pausing/restarting training is just lost revenue for the user... and that's not even guaranteed. I'd say it's pretty minimal. 

Super long term, we'd love to see the mine be useful for the user too, but I don't even really put it on my radar at the moment. It's not a part of this use case (that's more of a... given we're super successful, where could we pivot?)

**Comment by cereallarceny** [comment]
@iamtrask I'm actually not sure I agree about the lack of consistent Internet access.  In my mind, a mine should always be online and available to take new models from Sonar.  Part of that reasoning is lost funds, but it's also that data scientists need a swath of mines constantly online to train their models.  The more mines online, the quicker the model gets trained (in theory).  This is definitely true early on until we've amassed thousands or millions of mines.  I almost see it that we "owe it" to the data scientists to have mines online 24/7.  We're providing the service of federated machine learning to them, not unlike a public utility.  Instead of electricity, water, or natural gas, we provide access to data and computational power surrounding the use of that data.

**Comment by iamtrask** [comment]
@cereallarceny I think we're on the same page, as I agree with all of your reasons. In an perfect world, everyone's mines would be up all the time. Perhaps a better way to articulate my priorities is with a simple list.

1) Privacy: users are the only ones who can see their own data (thus cloud hosting we control/have access to is a non-starter)
2) Speed of Development: we need a mine that you and I can turn on before we need one that Mom and Dad can turn on (sort of... crawl before we run)
3) Ease of Deployment: people can turn on a desktop app or iOS app faster than they can spin up their own EC2 instance or buy a raspberry PI.
4) Uptime of Servers: in a perfect world, mines would be in the personal clouds of users.

Also, another key limitation is that the data must be in a **decrypted state** when in the mine. We cannot yet rely on data encryption because it prevents us from doing machine learning (it would require homomorphic encryption with two different public keys, one for the model and another for the data, and that tech doesn't really work yet to my knowledge)

**Comment by anoff** [comment]
Why are you opposed to having a central server with knowledge of the mines? Kind of like a Mine Index. It knows they are there and with optional ways of contacting the mine owners you reduce the chance of missed income for the Miner and at the same time reduce uptime requirements. Keeping a (non-cloud) Mine up 24/7 is lot of effort - unless you have the NAS solution Patrick listed in the beginning.
I think the priority is to focus on the Miners instead of the Data Scientists. The whole paradigm is that Data is more valuable than models. So we should make sure we get enough data. That attracts the data scientists. The other way around doesn't feel as natural to me.

Would it help if i drew up the ecosystem I'm imagining or are we on the same page when we talk about central server knowing about mines?

**Comment by iamtrask** [comment]
I'm not remotely opposed to a central server with knowledge of the mines. In fact, the good news there is that the blockchain itself (controlled by Sonar) is already a central server with knowledge of every mine. Each mine has a public key and a private key, and the public keys have to be registered on the blockchain (in Sonar) in order to receive funds from the Data Scientists.

**Comment by iamtrask** [comment]
I've updated the docs workflow if it helps to explain the role of the Mine. https://github.com/OpenMined/docs/tree/master/workflow
